### PR TITLE
[Metrics SDK] Metric aggregation temporality controls 

### LINKED ISSUE #1339: [Metrics SDK] AggregationTemporality Controls
Spec change: https://github.com/open-telemetry/opentelemetry-specification/pull/2404

The concept of "preferred temporality" is removed from the spec. Temporality is now supposed to be decided by the MetricReader which can take a callback which decides temporality.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1339

## Changes

As preferred aggregation temporality concept is removed from specs, and is made as function of instrument-kind (see section here - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader) , following changes are done -

- Modify `GetAggregationTemporality()` method in MetricReader. The methods now take in instrument type/kind, that lets exporter find which temporality to use based on the instrument.
- Add  `GetAggregationTemporality()` method in MetricExporter. This method takes the instrument type/kind, and returns the correct temporality to use based on the exporter used - 
    - OTLP exporter can be configured with a preferred temporality
    - Prometheus only uses cumulative (as it doesn't support delta).

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_grpc_metric_exporter_options.h [hunk 1]
```diff
@@ -19,8 +19,13 @@ namespace otlp
...
 struct OtlpGrpcMetricExporterOptions : public OtlpGrpcExporterOptions
 {
-  opentelemetry::sdk::metrics::AggregationTemporality aggregation_temporality =
-      opentelemetry::sdk::metrics::AggregationTemporality::kDelta;
+
+  // Preferred Aggregation Temporality
+  sdk::metrics::AggregationTemporality aggregation_temporality =
+      sdk::metrics::AggregationTemporality::kCumulative;
+
+  // opentelemetry::sdk::metrics::AggregationTemporalitySelector aggregation_temporality_selector =
```
Comment (ThomsonTan) [inline exporters/otlp/include/opentelemetry/exporters/otlp/otlp_grpc_metric_exporter_options.h:27]: What is this for?

File: exporters/otlp/src/otlp_metric_utils.cc [hunk 2]
```diff
@@ -231,6 +231,41 @@ void OtlpMetricUtils::PopulateRequest(
...
+    sdk::metrics::InstrumentType instrument_type) noexcept
+{
+  switch (instrument_type)
+  {
+    case sdk::metrics::InstrumentType::kCounter:
+    case sdk::metrics::InstrumentType::kObservableCounter:
+    case sdk::metrics::InstrumentType::kHistogram:
+    case sdk::metrics::InstrumentType::kObservableGauge:
+      return sdk::metrics::AggregationTemporality::kDelta;
+      break;
```
Comment (ThomsonTan) [inline exporters/otlp/src/otlp_metric_utils.cc:255]: Seems this break is redundant?

**Review by ThomsonTan — APPROVED** [review]

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_grpc_metric_exporter_options.h (hunk 1 above)
Comment (lalitb) [reply exporters/otlp/include/opentelemetry/exporters/otlp/otlp_grpc_metric_exporter_options.h]: Leftover from debugging, shouldn't be there. Removed thanks :)

File: exporters/otlp/src/otlp_metric_utils.cc (hunk 2 above)
Comment (lalitb) [reply exporters/otlp/src/otlp_metric_utils.cc]: Good catch. Removed break now :)

**Comment by ThomsonTan** [comment]
@lalitb can you please resolve the merging conflict?

**Comment by lalitb** [comment]
> @lalitb can you please resolve the merging conflict?

Fixed. Thanks.

**Pull request merged**